### DISCUSSION TITLE
Is it possible to get the "path" to a sub-model inside a field validator?

### DISCUSSION THREAD

**Discussion by gwerbin** [pydantic/pydantic#4724] [Question] [answered]
```python
import logging
from pydantic import BaseModel, ModelField

logger = logging.getLogger(__name__)

class Widget(BaseModel):
    height: float
    width: float

    @validator("height", "width"):
    def check_positive(cls, x: float, field: ModelField) -> x:
        logger.debug("Validating field: %r", field.name)
        if x <= 0:
            raise ValueError(f"{field.name} must be positive.")
        return x

class Bucket(BaseModel):
    widget_left: Widget
    widget_right: Widget
```

When creating an instance of `Bucket`, it would be very useful if my log message could indicate that the "path" to the `height` field in `widget_left` is something like `widget_left.height`.

However, I only see how to access `field.name` inside the validator, which is just `height`.

Is there some way to tap into the current validator "state" during instance creation/validation in order to see this? I suppose I could (very hackily) create a `@root_validator` that inserts a value into a global cache, but that seems extremely flaky and hard to do in a concurrency-safe way.

**Comment by samuelcolvin** [comment, marked as answer]
Sadly no it's not at the moment.

It's something we might be able to support in v2.

**Reply by mecampbellsoup** [reply]
Any update on this now that v2 is released?

**Reply by rmorshea** [reply]
I'm not seeing any clear way to accomplish this in v2. I could be wrong though.

**Comment by jmehnle** [comment]
In order to debug strange interactions between different validators (across different models), I could really, _really_ use a way to know the field path inside a validator. Sadly, this information is not made available to validators even in Pydantic v2. I wish there was a way …

**Comment by free-soellingeraj** [comment]
In order to do partial model validation, was looking for a solution like this, essentially to get a list of strings for a key path, as is supported by postgres sqlalchemy.  Is this planned?

See:
```python
    def _apply_jsonb_updates_from_partial(
        self, column, partial_update: dict, path: list = None
    ) -> ColumnElement:
        """Recursively apply partial updates using jsonb_set for specific paths.

        Args:
            column: The JSONB column to update
            partial_update: Partial update dictionary
            path: Current path in the JSON structure

        Returns:
            SQLAlchemy expression with applied updates
        """
        if path is None:
            path = []

        result_column = column

        for key, value in partial_update.items():
            current_path = path + [key]

            if isinstance(value, dict):
                # Recursively handle nested dictionaries for partial updates
                result_column = self._apply_jsonb_updates_from_partial(
                    result_column, value, current_path
                )
            else:
                # Convert enums to their string values for proper JSONB serialization
                # to_jsonb can't handle enums directly..
                if isinstance(value, Enum):  # Handle enum types
                    json_value = value.value
                else:
                    json_value = value

                # Set the value at the specific path
                result_column = func.jsonb_set(
                    result_column,
                    cast(current_path, ARRAY(Text)),
                    func.to_jsonb(json_value),
                    True,  # create_missing, here we've already validated that all keys are expected via DTO
                )
```
If I could get the flat key path version of the data structure I could eliminate the recursive loop here.
